bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java

```diff
@@ -12,34 +12,87 @@
 import org.openhab.core.ui.components.UIComponentRegistryFactory;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
 
 /**
- * Implementation for a {@link UIComponentRegistryFactory} using a {@link StorageService} and a
- * {@link UIComponentProvider}.
+ * Implementation for a {@link UIComponentRegistryFactory} using a set of {@link UIComponentProvider}.
  *
  * @author Yannick Schaus - Initial contribution
+ * @author Łukasz Dywicki - Removed explicit dependency on storage providers.
+ * @author Jonathan Gilbert - Made providers' collections immutable.
  */
 @Component(service = UIComponentRegistryFactory.class, immediate = true)
 public class UIComponentRegistryFactoryImpl implements UIComponentRegistryFactory {
-    Map<String, UIComponentRegistryImpl> registries = new HashMap<>();
-
-    @Reference
-    StorageService storageService;
+    Map<String, UIComponentRegistryImpl> registries = new ConcurrentHashMap<>();
+    Map<String, Set<UIProvider>> providers = new ConcurrentHashMap<>();
 
     @Override
     public UIComponentRegistryImpl getRegistry(String namespace) {
         UIComponentRegistryImpl registry = registries.get(namespace);
         if (registry == null) {
-            registry = new UIComponentRegistryImpl(namespace, storageService);
+            Set<UIProvider> namespaceProviders = this.providers.get(namespace);
+            ManagedProvider<@NonNull RootUIComponent, @NonNull String> managedProvider = null;
+            if (namespaceProviders != null) {
+                for (UIProvider provider : namespaceProviders) {
+                    if (provider instanceof ManagedProvider) {
+                        managedProvider = (ManagedProvider<@NonNull RootUIComponent, @NonNull String>) provider;
+                        break;
+                    }
+                }
+            }
+            registry = new UIComponentRegistryImpl(namespace, managedProvider, namespaceProviders);
             registries.put(namespace, registry);
         }
         return registry;
     }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    void addProvider(UIProvider provider) {
+        if (registries.containsKey(provider.getNamespace())) {
+            registries.get(provider.getNamespace()).addProvider(provider);
+        }
+        registerProvider(provider);
+    }
+
+    void removeProvider(UIProvider provider) {
+        if (registries.containsKey(provider.getNamespace())) {
+            registries.get(provider.getNamespace()).removeProvider(provider);
+        }
+        deregisterProvider(provider);
+    }
+
+    private void registerProvider(UIProvider provider) {
+        Set<UIProvider> existing = providers.get(provider.getNamespace());
```

Comment:
@splatch I modified this section to make the collections immutable, as I was having issues with them not being (not sure why).

Reply:
Making final set inserted into map immutable is wrong, cause you can't remove provider later on when it disappear. 

Given that component activation can be made through different threads (ie. declarative services, OSGi activator etc), you probably better of with synchronized collection rather than plain `HashSet`. Keep in mind that set registered for namespace will be manipulated.

Reply:
Well, whilst the set was immutable, I replaced it with a completely new set when it was modified, so it did still work.

Saying this, I don't really like the verbose code to do this (Java mutable collections I guess). Instead I have now changed it to use a `CopyOnWriteArraySet`, and also synchronized write access (otherwise concurrent updates could lose elements), as it should happen very infrequently anyway.